data_model/derive/src/event_set.rs

```diff
@@ -0,0 +1,384 @@
+
+                    let mut need_comma = false;
+
+                    #(if self.contains(#flag_idents) {
+                        if need_comma {
+                            write!(f, ", ")?;
+                        } else {
+                            need_comma = true;
+                        }
+                        write!(f, "{}", #flag_names)?
+                    })*
+
+                    write!(f, "]")
+                }
+            }
+
+            impl core::ops::BitOr for #set_ident {
+                type Output = Self;
+
+                fn bitor(self, rhs: Self) -> Self {
+                    self.or(rhs)
+                }
+            }
+
+            impl core::ops::BitOrAssign for #set_ident {
+                fn bitor_assign(&mut self, rhs: Self) {
+                    *self = self.or(rhs);
+                }
+            }
+
+            impl serde::Serialize for #set_ident {
+                fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+                where
+                    S: serde::Serializer,
+                {
+                    use serde::ser::*;
+
+                    let basis = self.decompose();
+                    let mut seq = serializer.serialize_seq(Some(basis.len()))?;
+                    for event in basis {
+                        let str = if false {
+                            "dummy"
+                        } #(else if event.contains(#flag_idents) {
+                            #flag_names
+                        })* else {
+                            unreachable!();
+                        };
+
+                        seq.serialize_element(&str)?;
+                    }
+                    seq.end()
+                }
+            }
+
+            impl<'de> serde::Deserialize<'de> for #set_ident {
+                fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+                where
+                    D: serde::Deserializer<'de>,
+                {
+                    struct Visitor;
+
+                    impl<'de> serde::de::Visitor<'de> for Visitor {
+                        type Value = #set_ident;
+
+                        fn expecting(&self, formatter: &mut core::fmt::Formatter) -> core::fmt::Result {
+                            formatter.write_str("a sequence of strings")
+                        }
+
+                        fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+                        where
+                            A: serde::de::SeqAccess<'de>,
+                        {
+                            let mut result = #set_ident::empty();
+
+                            // can't use Cow directly because we need to refer to it by different names depending on whether we're in std or not
+                            while let Some(event) = seq.next_element::<#krate::derive_private::Cow<'_, str>>()? {
+                                match core::convert::AsRef::as_ref(&event) {
```

Comment:
This should work just fine: 

```suggestion
                            while let Some(event) = seq.next_element::<&'de str>()? {
                                match event {
```

Reply:
Hm, I used `&str`, and it didn't work at all. It's supposed to work in most cases, except when there are escape sequences in the string (which depends on the serialization format), but obviously no escapes are needed for the event names.. Maybe if I have used `&'de str` id would have actually worked (does `&str` actually get treated as `&'static str`?).

I think I still prefer the `Cow` here, because it will always work and will not give a confusing error message when a string copy is needed by the serialization format...

Reply:
Hmm, how `Cow` can help with escape sequences in strings?

> does &str actually get treated as &'static str?

Doesn't look like this.
Because when i used either `&str or &'de str` everything compiled just fine, but when used `&'static str` rust complained about lifetimes.

I don't understend why `Cow` would work in places where `&str` wouldn't.

Default `serde` implementation provides a way to deserialize `&'a str` from any `Deserialzier` by `deserialize_str` + for `Deserializer` it's required to implement `deserialize_str` method. 

Only if implementator of `deserialize_str` would throw a panic.

Looking at implementation of `Deserialize` for `Cow<'a, T>` i can see that you can just use `String` instead.

Reply:
It's a bit complicated, but serde can pass a string to your visitor in three ways:
- [`visit_str`](https://docs.rs/serde/latest/serde/de/trait.Visitor.html#method.visit_str): gives you a `&str` with function-local lifetime (it can't escape it)
- [`visit_borrowed_str`](https://docs.rs/serde/latest/serde/de/trait.Visitor.html#method.visit_borrowed_str): gives you a `&'de str`, a string _can_ escape the function and be stored in the deserialized structure
- [`visit_string`](https://docs.rs/serde/latest/serde/de/trait.Visitor.html#method.visit_string): gives you an owned string value

`visit_str` is the most general, both `visit_string` and `visit_borrowed_str` delegate to it by default.

> how `Cow` can help with escape sequences in strings?

In serialization formats needing escaping, a borrowed string can not always be produced, as the string would need rewriting (no way to get a reference to "a\nb" from string "a\\nb"), so an allocation is required.

The only implementation of [`Deserialize` for `&str`](https://docs.rs/serde/latest/src/serde/de/impls.rs.html#621-641) only accepts `visit_borrowed_str`, so it will work only without escapes.

In my mind Cow was supposed to efficiently handle both the owned and borrowed cases, but in actuality it doesn't. I'll write a custom newtype that will have a visitor implementing the most general `visit_str` function then.

Reply:
> In serialization formats needing escaping, a borrowed string can not always be produced, as the string would need rewriting (no way to get a reference to "a\nb" from string "a\\nb"), so an allocation is required.

Aha, thanks for clarification

Reply:
pushed a change